File: src/execution/runtimeSandboxEngine.ts
/**
 * Converted from Python: core/execution/runtime_sandbox_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";

export function buildRuntimeSandbox(runtime: any = "browser", allowed_actions: any = null, rollback_enabled: any = true, max_actions: any = 1000, timeout_ticks: any = 10000, replay_policy: any = "strict"): any {
  var default_allowed: any = ["browser_click", "browser_focus", "native_focus"];
  if (py.eq(runtime, "terminal")) {
    default_allowed = ["terminal_command"];
  } else if (py.eq(runtime, "native")) {
    default_allowed = ["native_focus"];
  } else if (py.eq(runtime, "vm")) {
    default_allowed = ["vm_execute"];
  }
  return {"runtime": runtime, "allowed_actions": py.sorted(py.or2(allowed_actions, () => (default_allowed))), "rollback_enabled": rollback_enabled, "max_actions": max_actions, "execution_boundaries": {"max_depth": 50, "max_mutations": 100, "timeout_ticks": timeout_ticks}, "mutation_limits": {"max_mutations": 100}, "rollback_policy": {"enabled": rollback_enabled}, "timeout_policy": {"ticks": timeout_ticks}, "replay_policy": replay_policy, "bounded": true};
}

File: lib/src/execution/runtime_sandbox_engine.dart
/// Port of core/execution/runtime_sandbox_engine.build_runtime_sandbox.
Map<String, dynamic> buildRuntimeSandbox({
  String runtime = 'browser',
  List<String>? allowedActions,
  bool rollbackEnabled = true,
  int maxActions = 1000,
  int timeoutTicks = 10000,
  String replayPolicy = 'strict',
}) {
  List<String> defaultAllowed = <String>[
    'browser_click',
    'browser_focus',
    'native_focus',
  ];
  if (runtime == 'terminal') {
    defaultAllowed = <String>['terminal_command'];
  } else if (runtime == 'native') {
    defaultAllowed = <String>['native_focus'];
  } else if (runtime == 'vm') {
    defaultAllowed = <String>['vm_execute'];
  }

  final List<String> allowed = <String>[...(allowedActions ?? defaultAllowed)]
    ..sort();

  return <String, dynamic>{
    'runtime': runtime,
    'allowed_actions': allowed,
    'rollback_enabled': rollbackEnabled,
    'max_actions': maxActions,
    'execution_boundaries': <String, dynamic>{
      'max_depth': 50,
      'max_mutations': 100,
      'timeout_ticks': timeoutTicks,
    },
    'mutation_limits': <String, dynamic>{'max_mutations': 100},
    'rollback_policy': <String, dynamic>{'enabled': rollbackEnabled},
    'timeout_policy': <String, dynamic>{'ticks': timeoutTicks},
    'replay_policy': replayPolicy,
    'bounded': true,
  };
}
